### PR TITLE
Prioritize packet events over local events

### PR DESCRIPTION
**Pull request by stevenengler**
Previously, when the host would get the next event and there were multiple events at the same time, it would prioritize events with a lower source host ID. This means that sometimes packets would be prioritized over local events, and sometimes local events would be prioritized depending on the source host's ID relative to the current host's ID. This PR removes this inconsistency, and now packet events will always be prioritized over local events.

This PR also splits the event into separate event types: packet events (that contain a packet object) and local events (that contain a task). This means that we can take a packet object directly from a packet event, which is more flexible than running an arbitrary task closure containing the packet object.

### CODE REVIEW

File: src/main/core/work/event.rs [hunk 1]
```diff
@@ -8,51 +9,53 @@ use super::task::TaskRef;
...
-    src_host_id: HostId,
-    dst_host_id: HostId,
-    src_host_event_id: u64,
+    data: EventData,
     _counter: ObjectCounter,
 }
 
 impl Event {
-    pub fn new(task: TaskRef, time: EmulatedTime, src_host: &Host, dst_host_id: HostId) -> Self {
+    /// A new packet event.
```
Comment (robgjansen) [inline src/main/core/work/event.rs:18]: Expand the comment a bit to explain what a packet event is.

File: src/main/core/work/event.rs [hunk 2]
```diff
@@ -8,51 +9,53 @@ use super::task::TaskRef;
...
-    pub fn execute(self, host: &Host) {
-        self.magic.debug_check();
-
-        // make sure we're executing on the correct host
-        assert_eq!(self.host_id(), host.id());
-
-        host.continue_execution_timer();
-        self.task.execute(host);
-        host.stop_execution_timer();
+    /// A new local event.
```
Comment (robgjansen) [inline src/main/core/work/event.rs:32]: Expand the comment a bit to explain what a local event is.

**Review by robgjansen — APPROVED** [review]

File: src/main/core/work/event.rs (hunk 1 above)
Comment (stevenengler) [reply src/main/core/work/event.rs]: Added.

> A new packet event, which is an event for packets arriving from the Internet. Packet events do not include packets on localhost.

File: src/main/core/work/event.rs (hunk 2 above)
Comment (stevenengler) [reply src/main/core/work/event.rs]: Added.

> A new local event, which is an event that was generated locally by the host itself (timers, localhost packets, etc).

**Comment by stevenengler** [comment]
The network and simulation performance is significantly different with this change: https://github.com/shadow/benchmark-results/tree/master/tor/2022-11-07-T14-56-21

@robgjansen Any issues with this before I merge it?

run_time

transfer_time_1048576 exit

**Comment by stevenengler** [comment]
We compared a benchmark run of this PR ([2022-11-14-T17-46-27](https://github.com/shadow/benchmark-results/tree/master/tor/2022-11-14-T17-46-27)) and the previous weekly benchmark run ([2022-11-12-T03-50-23](https://github.com/shadow/benchmark-results/tree/master/tor/2022-11-12-T03-50-23)) using the host heartbeat messages and the scripts in `src/tools`, and didn't find the [results](https://github.com/shadow/shadow/files/10042550/shadow.results.pdf) to be significantly different.

We have found that the number of allocations and syscalls were significantly less in the PR benchmark run.

PR:

```
Global syscall counts: {read:744607152, epoll_wait:390993063, write:299110729, epoll_ctl:151463882, getsockopt:83843256, ioctl:82845185, sendto:74909883, recvfrom:53884213, timerfd_settime:43373158, getrandom:25515138, close:8878698, lseek:7493425, open:5025897, fstat:4628634, openat:2746635, timerfd_create:2681458, setsockopt:1999664, socket:1893475, connect:1891057, getsockname:1810735, uname:1769203, fcntl:1765211, accept:1757246, shutdown:1545519, futex:1287931, accept4:999312, brk:304482, mmap:16699, shadow_hostname_to_addr_ipv4:16651, rt_sigaction:12176, munmap:12156, mprotect:7713, getpid:2180, shadow_init_memory_manager:2180, bind:2061, listen:2061, epoll_create:1861, fchmod:1190, epoll_create1:833, flock:833, pipe2:833, sysinfo:833, clone:714, getdents64:714, rt_sigprocmask:714, set_robust_list:714, shadow_get_shm_blk:714, eventfd2:357}
Global allocated object counts: {Packet:2028866038, TaskRef:1957435879, Event:1851421582, Payload:973770742, StatusListener:220451349, SysCallCondition:212311463, Timer:162451727, LegacyDescriptor:14232794, RegularFile:7779072, TCP:3766319, TimerFd:2681458, Futex:212057, Epoll:5588, ManagedThread:2894, SysCallHandler:2894, Process:2180, NetworkInterface:1742, FutexTable:871, HostCInternal:871, Router:871, UDP:357}
```

Weekly of main branch:

```
Global syscall counts: {read:1232331835, epoll_wait:795098338, write:302624449, epoll_ctl:255307445, recvfrom:197039394, getsockopt:85155365, ioctl:84157468, sendto:82957731, timerfd_settime:44315297, getrandom:25595377, close:8885211, lseek:7488831, open:5040479, fstat:4628808, openat:2746662, timerfd_create:2681484, setsockopt:1999283, socket:1893294, connect:1890876, getsockname:1810636, uname:1769222, fcntl:1765249, accept:1757359, shutdown:1545604, futex:1286445, accept4:999112, brk:299065, mmap:16833, shadow_hostname_to_addr_ipv4:16657, munmap:12277, rt_sigaction:12176, mprotect:7510, getpid:2180, shadow_init_memory_manager:2180, bind:2061, listen:2061, epoll_create:1861, fchmod:1190, epoll_create1:833, flock:833, pipe2:833, sysinfo:833, clone:714, getdents64:714, rt_sigprocmask:714, set_robust_list:714, shadow_get_shm_blk:714, eventfd2:357}
Global allocated object counts: {TaskRef:4697115261, Event:2593379194, Packet:2055394681, Payload:989337824, StatusListener:580971615, SysCallCondition:572748357, Timer:521813618, LegacyDescriptor:14247062, RegularFile:7793681, TCP:3765952, TimerFd:2681484, Futex:211922, Epoll:5588, ManagedThread:2894, SysCallHandler:2894, Process:2180, NetworkInterface:1742, FutexTable:871, HostCInternal:871, Router:871, UDP:357}
```

**Comment by robgjansen** [comment]
With the new TGen benchmark, no change in simulated network performance:

Time to first byte

Time to last byte

But we do still get the [improvement in run time](https://github.com/shadow/benchmark-results/blob/master/tgen/2022-11-24-T15-01-00/plots/shadow.results.pdf).

**Comment by robgjansen** [comment]
@stevenengler please go ahead and merge at your convenience :)

**Comment by stevenengler** [comment]
Rebased and re-ran the benchmarks since a lot has changed since this PR was opened (such as the new relay and router code). The run-time performance improves significantly, and the network performance changes slightly.

### [Tor](https://github.com/shadow/benchmark-results/tree/master/tor/2023-02-21-T03-13-52)

1676991147_grim
1676991211_grim

### [Tgen](https://github.com/shadow/benchmark-results/tree/master/tgen/2023-02-21-T00-27-43)

1676991043_grim
1676991091_grim

**Comment by stevenengler** [comment]
@robgjansen Re-marking you as reviewer. You've already reviewed the code once, but it's been a while so letting you have another look if you want.

File: src/main/core/work/event.rs [hunk 3]
```diff
@@ -61,105 +66,121 @@ impl PartialEq for Event {
...
+    task: TaskRef,
+    event_id: u64,
+}
+
+impl PacketEventData {
+    /// Get the packet contained within this event.
+    pub fn packet(self) -> Packet {
+        self.packet
+    }
+}
```
Comment (robgjansen) [inline src/main/core/work/event.rs:130]: Do we want to do this by implementing the `From` trait instead?

File: src/main/core/work/event.rs [hunk 4]
```diff
@@ -61,105 +66,121 @@ impl PartialEq for Event {
...
-        let event = unsafe { Box::from_raw(event) };
-        let host = unsafe { host.as_ref().unwrap() };
-
-        event.execute(host);
+impl LocalEventData {
+    /// Get the task contained within this event.
+    pub fn task(self) -> TaskRef {
+        self.task
     }
+}
```
Comment (robgjansen) [inline src/main/core/work/event.rs:163]: Implement the `From` trait here instead?

**Review by robgjansen — APPROVED** [review]
Here is my understanding of why we see the runtime performance improvement with this PR.

Prior to this PR, sometimes (depending on the host id) local events would be prioritized ahead of packet events that happen at the same instant, meaning that we could move a single packet event from the event queue to the router codel queue, which would cause a new local event to start the relay forward loop, which would get placed in front of other waiting packets events in the event queue. Effectively we would create a new relay forward event for every packet, and run the event to forward only one packet, and then repeat when the next packet gets added to the router codel queue again.

With this PR, all packets are consistently getting moved from the event queue to the router codel queue in one batch, and then after that we run the relay forward loop which can then forward as many packets as possible. This significantly reduces the number of event objects (and related push and pop operations) required to receive a set of packets that arrive at the same instant.

File: src/main/core/work/event.rs (hunk 3 above)
Comment (stevenengler) [reply src/main/core/work/event.rs]: Sure, I've made it `impl From<PacketEventData> for Packet` instead.

File: src/main/core/work/event.rs (hunk 4 above)
Comment (stevenengler) [reply src/main/core/work/event.rs]: I've made it `impl From<LocalEventData> for TaskRef` instead.

**Pull request merged**